File: backend/tts-server.js
/**
 * Edge-TTS Backend Server
 * Free text-to-speech using Microsoft Edge TTS
 * 
 * Endpoints:
 *   GET  /api/tts/voices    - List all available voices grouped by language
 *   POST /api/tts/synthesize - Generate audio from text
 *   POST /api/tts/preview   - Generate short voice preview
 */

import express from 'express';
import cors from 'cors';
import { EdgeTTS, VoicesManager } from 'edge-tts-universal';

const app = express();
const PORT = 5050;

// CORS
app.use(cors({
    origin: ['http://localhost:3000', 'http://localhost:5173'],
    credentials: true,
}));
app.use(express.json({ limit: '10mb' }));

// Cache voices
let voicesCache = null;

/**
 * GET /api/tts/voices
 * Returns all voices grouped by language
 */
app.get('/api/tts/voices', async (req, res) => {
    try {
        if (!voicesCache) {
            console.log('🔄 Fetching voices from Microsoft Edge TTS...');
            const manager = await VoicesManager.create();
            const allVoices = manager.voices;

            // Group by locale
            const grouped = {};
            for (const v of allVoices) {
                const locale = v.Locale;
                const langName = v.Locale;
                // Try to extract a friendly name from FriendlyName
                const friendlyLabel = v.FriendlyName || locale;

                if (!grouped[locale]) {
                    grouped[locale] = {
                        locale,
                        language: friendlyLabel.split(' - ')[0] || locale,
                        voices: [],
                    };
                }

                // Clean display name
                let displayName = v.FriendlyName || v.ShortName;
                displayName = displayName
                    .replace('Microsoft Server Speech Text to Speech Voice ', '')
                    .replace(`(${v.Locale}, `, '(')
                    .trim();

                grouped[locale].voices.push({
                    name: v.ShortName,
                    displayName,
                    gender: v.Gender,
                    locale: v.Locale,
                });
            }

            // Sort by locale, move en-US to top
            let result = Object.values(grouped).sort((a, b) =>
                a.locale.localeCompare(b.locale)
            );

            const enUS = result.find(l => l.locale === 'en-US');
            if (enUS) {
                result = [enUS, ...result.filter(l => l.locale !== 'en-US')];
            }

            voicesCache = result;
            console.log(`✅ Cached ${allVoices.length} voices in ${result.length} languages`);
        }

        res.json({ voices: voicesCache });
    } catch (error) {
        console.error('❌ Error listing voices:', error);
        res.status(500).json({ error: error.message });
    }
});

/**
 * POST /api/tts/synthesize
 * Body: { text, voice, rate }
 * Returns: audio/mpeg blob
 */
app.post('/api/tts/synthesize', async (req, res) => {
    const { text, voice = 'en-US-GuyNeural', rate = '+0%' } = req.body;

    if (!text || !text.trim()) {
        return res.status(400).json({ error: 'Text cannot be empty' });
    }

    console.log(`🎤 Synthesizing: voice=${voice}, rate=${rate}, text=${text.length} chars`);

    try {
        const tts = new EdgeTTS(text, voice, { rate });
        const result = await tts.synthesize();

        // Convert Blob/ArrayBuffer to Buffer
        const arrayBuffer = await result.audio.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);

        console.log(`✅ Audio generated: ${buffer.length} bytes`);

        res.set({
            'Content-Type': 'audio/mpeg',
            'Content-Length': buffer.length,
            'Content-Disposition': 'attachment; filename=voiceover.mp3',
        });
        res.send(buffer);
    } catch (error) {
        console.error('❌ Synthesis error:', error);
        res.status(500).json({ error: error.message });
    }
});

/**
 * POST /api/tts/preview
 * Body: { voice }
 * Returns: short audio preview
 */
app.post('/api/tts/preview', async (req, res) => {
    const { voice = 'en-US-GuyNeural' } = req.body;
    const previewText = 'Hello! This is a preview of my voice. I hope you like how I sound.';

    console.log(`🔊 Preview: voice=${voice}`);

    try {
        const tts = new EdgeTTS(previewText, voice);
        const result = await tts.synthesize();

        const arrayBuffer = await result.audio.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);

        res.set({
            'Content-Type': 'audio/mpeg',
            'Content-Length': buffer.length,
        });
        res.send(buffer);
    } catch (error) {
        console.error('❌ Preview error:', error);
        res.status(500).json({ error: error.message });
    }
});

/**
 * Health check
 */
app.get('/api/tts/health', (req, res) => {
    res.json({ status: 'ok', service: 'Edge-TTS Server' });
});

app.listen(PORT, () => {
    console.log(`
${'='.repeat(50)}
🎤 Edge-TTS Server
${'='.repeat(50)}

📡 Server: http://localhost:${PORT}
🏥 Health: http://localhost:${PORT}/api/tts/health

${'='.repeat(50)}
    `);
});

File: backend/tts_server.py
"""
Edge-TTS Server
Free text-to-speech using Microsoft Edge TTS
Replaces ElevenLabs API

Endpoints:
  GET  /api/tts/voices    - List all available voices grouped by language
  POST /api/tts/synthesize - Generate audio from text
  POST /api/tts/preview   - Generate short voice preview
"""

import asyncio
import os
import uuid
import edge_tts
from fastapi import FastAPI, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import FileResponse
from pydantic import BaseModel
from typing import Optional

app = FastAPI(title="Edge-TTS Server", version="1.0.0")

# CORS
app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:3000", "http://localhost:5173"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# Audio output directory
AUDIO_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "public", "temp", "audio")
os.makedirs(AUDIO_DIR, exist_ok=True)

# Cache voices list
_voices_cache = None


class SynthesizeRequest(BaseModel):
    text: str
    voice: str = "en-US-GuyNeural"
    rate: str = "+0%"


class PreviewRequest(BaseModel):
    voice: str = "en-US-GuyNeural"


async def get_all_voices():
    """Fetch and cache all edge-tts voices"""
    global _voices_cache
    if _voices_cache is not None:
        return _voices_cache

    voices = await edge_tts.list_voices()

    # Group by language
    grouped = {}
    for v in voices:
        locale = v["Locale"]
        lang_name = v.get("LocaleName", locale)

        if locale not in grouped:
            grouped[locale] = {
                "locale": locale,
                "language": lang_name,
                "voices": []
            }

        grouped[locale]["voices"].append({
            "name": v["ShortName"],
            "displayName": v["FriendlyName"].replace("Microsoft Server Speech Text to Speech Voice ", "").replace(f"({v['Locale']}, ", "(").rstrip(")") + ")",
            "gender": v["Gender"],
            "locale": locale,
        })

    # Sort by language name
    result = sorted(grouped.values(), key=lambda x: x["language"])

    # Move English-US to top
    en_us = next((l for l in result if l["locale"] == "en-US"), None)
    if en_us:
        result.remove(en_us)
        result.insert(0, en_us)

    _voices_cache = result
    return result


@app.get("/api/tts/voices")
async def list_voices():
    """Get all available voices grouped by language"""
    try:
        voices = await get_all_voices()
        return {"voices": voices}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@app.post("/api/tts/synthesize")
async def synthesize(req: SynthesizeRequest):
    """Generate audio from text"""
    if not req.text.strip():
        raise HTTPException(status_code=400, detail="Text cannot be empty")

    filename = f"{uuid.uuid4()}.mp3"
    filepath = os.path.join(AUDIO_DIR, filename)

    try:
        communicate = edge_tts.Communicate(
            text=req.text,
            voice=req.voice,
            rate=req.rate,
        )
        await communicate.save(filepath)

        return FileResponse(
            filepath,
            media_type="audio/mpeg",
            filename=filename,
            headers={
                "Content-Disposition": f"attachment; filename={filename}",
                "X-Audio-Filename": filename,
            }
        )
    except Exception as e:
        if os.path.exists(filepath):
            os.remove(filepath)
        raise HTTPException(status_code=500, detail=str(e))


@app.post("/api/tts/preview")
async def preview(req: PreviewRequest):
    """Generate a short preview of a voice"""
    preview_text = "Hello! This is a preview of my voice. I hope you like how I sound."
    filename = f"preview_{req.voice.replace('-', '_')}_{uuid.uuid4().hex[:8]}.mp3"
    filepath = os.path.join(AUDIO_DIR, filename)

    try:
        communicate = edge_tts.Communicate(
            text=preview_text,
            voice=req.voice,
        )
        await communicate.save(filepath)

        return FileResponse(
            filepath,
            media_type="audio/mpeg",
            filename=filename,
        )
    except Exception as e:
        if os.path.exists(filepath):
            os.remove(filepath)
        raise HTTPException(status_code=500, detail=str(e))


@app.get("/api/tts/health")
async def health():
    return {"status": "ok", "service": "Edge-TTS Server"}


if __name__ == "__main__":
    import uvicorn
    port = 5050
    print(f"""
    {'='*50}
    🎤 Edge-TTS Server
    {'='*50}

    📡 Server: http://localhost:{port}
    📚 Docs:   http://localhost:{port}/docs
    🏥 Health: http://localhost:{port}/api/tts/health

    {'='*50}
    """)
    uvicorn.run("tts_server:app", host="0.0.0.0", port=port, reload=True)
